### DISCUSSION TITLE
how to handle the tuple  longer than 36 or higher?

### DISCUSSION THREAD

**Discussion by wujilingfeng** [AccelerateHS/accelerate#520] [Q&A] [answered]
I need to handle longer arrays, and I found that the tuple didn't support it. Even if the tuple can support it,, It would be too much trouble to do so. 
it seems that Exp [Float] is better to do it, But it's also unsupported.

Acc vector seems  inappropriate，Because many functions need parameters like `Exp a->Exp b`

**Comment by tmcdonell** [comment, marked as answer]
The idea is that you have arrays of data (in `Acc`) and then you perform some (parallel) operation over the elements in those arrays, where the functions on the individual elements are of that `Exp` type.

If you think about the `map` function for example, the array data is stored in `Acc`, and the operation to apply to each element of the array is (as you saw) a function `Exp a -> Exp b`.

But maybe if you tell us a bit more about what you are trying to do we can give some more specific/helpful advice?

**Reply by wujilingfeng** [reply]
> The idea is that you have arrays of data (in `Acc`) and then you perform some (parallel) operation over the elements in those arrays, where the functions on the individual elements are of that `Exp` type.
> 
> If you think about the `map` function for example, the array data is stored in `Acc`, and the operation to apply to each element of the array is (as you saw) a function `Exp a -> Exp b`.
> 
> But maybe if you tell us a bit more about what you are trying to do we can give some more specific/helpful advice?

As you said, I'm using the map function.

I want to use exp float - > exp [float], because the result I want is an array of indefinite length

**Reply by dpvanbalen** [reply]
We do not support arrays or lists on the element level, but perhaps the function `expand` https://github.com/AccelerateHS/accelerate/blob/5971c5d8e4dbba28d2017e7ce422cf46a20197cb/src/Data/Array/Accelerate/Prelude.hs#L2516 covers what you need? 
Otherwise, could you zoom out a bit more so we can understand why you want the `Exp Float -> Exp [Float]` function?

**Reply by wujilingfeng** [reply]
I was wirting sift algorithm, every feature point has many directions.

**Reply by tmcdonell** [reply]
But in your implementation is "many" a statically known number, or does it vary per feature?

**Reply by wujilingfeng** [reply]
> But in your implementation is "many" a statically known number, or does it vary per feature?

It vary per feature.

**Reply by tmcdonell** [reply]
Maybe the `expand` function that David mentioned would work for you then. This is getting into the realm of irregular data-parallelism, which is much more difficult to execute in parallel... I'm interested to know what other features could help here!

(To answer your earlier question, one concrete reason we don't support something like `Exp [Float]` is simply that individual threads on the GPU can not allocate memory, so there is no way to build that list dynamically. Even if they could, then there is the problem of where in memory the (thousands of) threads should decide (in parallel) to store their results.)

**Reply by wujilingfeng** [reply]
Thank you! I think there are many indirect ways to solve it .
